### PR TITLE
lookup_service: add ability to assign and return unique ids

### PR DESCRIPTION
**Pull request by rohgarg**

### CODE REVIEW

**Comment by rohgarg** [comment]
@jiajuncao: Could you please check if this works for you?

The new API I have added can create a new NS database, generate a unique id, populate the database with the unique id, and return the generated unique id. This allows you to use the API first at launch time, and then leverage the regular NS API to do whatever you want later.

To use this API, you need to specify the following parameters:
 - a NS database name;
 - a key, which could be a host-id (if you wanted per-host unique ids, e.g., lids for IB), or a pid (if you wanted per process unique ids, e.g., QP numbers for IB);
 - key_len, which is the length of the key (in bytes);
 - val, which is the return value, (in other words, the unique id generated by the coordinator);
 - offset, which determines difference between two ids (e.g., a difference of 10 between two lids); and
 - val_len, which is the max. length in bytes for the unique id (e.g., 2-bytes for an lid).

File: src/lookup_service.cpp [hunk 1]
```diff
@@ -122,3 +129,32 @@ LookupService::respondToQuery(jalib::JSocket &remote,
...
+    }
+    JTRACE("Assigning a new unique id to client request")
+       (id) (_lastUniqueIds[id]);
+    KeyValue *v = new KeyValue(&_lastUniqueIds[id], val_len);
+    _lastUniqueIds[id] += _offsets[id];
+    kvmap[k] = v;
+  }
+
+  KeyValue *v = kvmap[k];
+  *val = new char[v->len()];
```
Comment (jiajuncao) [inline src/lookup_service.cpp]: Shall we add an assert here if `v->len() != val_len` ?

File: include/dmtcp.h.in [hunk 2]
```diff
@@ -211,6 +211,13 @@ EXTERNC int dmtcp_send_query_to_coordinator(const char *id,
                                             void *val,
                                             uint32_t *val_len);
```
Comment (jiajuncao) [inline include/dmtcp.h.in]: Can you add some comments about the function, especially the use cases?

File: src/coordinatorapi.cpp [hunk 3]
```diff
@@ -809,6 +809,57 @@ sendQueryToCoordinator(const char *id,
...
+  DmtcpMessage msg(DMT_NAME_SERVICE_GET_UNIQUE_ID);
+
+  JWARNING(strlen(id) < sizeof(msg.nsid));
+  strncpy(msg.nsid, id, sizeof msg.nsid);
+  msg.keyLen = key_len;
+  msg.valLen = 0;
+  msg.extraBytes = key_len;
+  msg.uniqueIdOffset = offset;
+  msg.valLen = *val_len;
+  int sock = coordinatorSocket;
```
Comment (jiajuncao) [inline src/coordinatorapi.cpp]: `sock` should be of type `jalib::JSocket`

**Pull request merged**